깃 허브 로그인 전략 및 콜백 함수 생성 및 프론트 라우팅

rolled-potatoes opened the pull request:
- [x] 깃 허브 로그인 전략 정의
- [x] 깃 허브 로그인 콜백 함수 정의
- [x] 깃 허브 로그인에서 넘어온 데이터를 저장
- [x] 프론트에서 깃 허브 로그인 후 token 을 저장
- [x] 로그인 페이지 생성
- [x] 로그인 페이지 라우팅

rolled-potatoes committed:
Temp (#74)

* Feat : GitHub passport 설정

깃 허브 oauth 로그인을 위한 전략 설정.

* Feat : jwt 생성 라이브러리 작성

* Feat : GitHub oauth callback 작성

깃 허브 로그인 성공 시 토큰을 받아 반환하는 콜백 함수 작성 함.

changheedev committed:
로그인 후 code를 token으로 요청하는 페이지 작업했습니다. (#76)

* Feat: GitHub 로그인 콜백 처리 페이지 컴포넌트 구현

깃허브 로그인의 콜백으로 code를 받아서 jwt 토큰을 요청하는 작업을 진행하는
페이지 컴포넌트 구현.

* Fix: 잘못된 문법 사용 코드 수정

Map.set()으로 사용해야하는 코드를 put() 으로 사용한 코드를 수정

* Fix: DropDown.js 파일내 잘못된 import path 수정

yejineee committed:
Feat: Login 화면 컴포넌트 구현 및 프론트 API_BASE_URL 환경 변수 추가 (#75)

* Chore: 프론트 웹팩 환경 변수로 API_BASE_URL추가

production 모드인지 아닌지에 따라서 api 요청 주소가 달라지게 된다.
api 요청 주소가 여러 곳에서 쓰이므로, 프론트의 환경 변수로 설정한다.

* Feat: 로그인 페이지 구현

첫 화면에 쓰이는 로그인 페이지 컴포넌트를 구현하였다.
버튼을 누르면 깃헙 로그인으로 넘어가게 된다.

Co-authored-by: yejineee <21700431@handong.edu>

changheedev committed:
Fix: 코드 공유간 누락된 코드 수정 (#77)

코드를 공유하면서 서로 누락된 부분을 다시 수정했습니다.

rolled-potatoes committed:
프론트 라우팅 설정 & 사용자 정보 관리 Store 생성  (#78)

* Feat : 사용자 정보 관리 store 생성

사용자 정보를 전역으로 접근 할 수 있는 store를 생성 함.

* Feat : 프론트 라우팅 생성

* Feat : 프론트 oauth callback 함수 작성

* Feat : 프론트 라우팅 분리 및 user store 재정의

프론트에서 특정 URL 제외 하기위해 하드코딩 한 것을 필요 따라서 렌더링 될 수 있게
라우팅을 서브 라우팅으로 분리 하였음.

기존 BaseComponent가 적절한 네이밍 아니라고 판단하여 변경 함.

changheedev commented on front/src/App.js:
```diff
@@ -17,13 +22,32 @@ const GlobalStyled = createGlobalStyle`
   }
 `;
 
-const App = () => {
+const TempComponent = () => <div>temp</div>;
```
테스트용으로 임시로 사용하기 위해 추가 하셨던 것 같은데 포함되지 않아야 할 것 같습니다.

rolled-potatoes replied:
그러게요. app.js 돌려 놓는다는걸 잊어버렸네요

changheedev commented on front/src/App.js:
```diff
@@ -17,13 +22,32 @@ const GlobalStyled = createGlobalStyle`
 
-const App = () => {
+const TempComponent = () => <div>temp</div>;
+const Template = () => {
   return (
     <>
```
여기도 임시로 Template이란 이름을 붙이신 것 같은데 인증이 필요한 라우터라는 것을 표현할 수 있는 이름으로 어떤게 좋을까요..??

yejineee replied:
어렵네요...authTemplate으로 하기엔 왠지 로그인 페이지 같은 느낌이구...

changheedev commented on front/src/App.js:
```diff
@@ -17,13 +22,32 @@ const GlobalStyled = createGlobalStyle`
-const App = () => {
+const TempComponent = () => <div>temp</div>;
+const Template = () => {
   return (
     <>
-      <GlobalStyled />
       <Header />
-      <IssueListPage />
+      <Link to="login">로그인</Link>
+      <Switch>
```
여기 로그인 Link 컴포넌트는 이슈 리스트 페이지에서 보였던 로그인 버튼인가요??

changheedev commented on front/src/routes/Login/LoginPage.js:
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import styled from 'styled-components';
+import color from '../../libs/color';
+
+const GithubButton = styled.button`
+  display: flex;
```
GitHub...ㅠㅠ

changheedev commented on front/src/routes/oauth/callback.js:
```diff
@@ -0,0 +1,20 @@
+  const history = useHistory();
+  const getToken = async () => {
+    const { code } = queryString.parse(location.search);
+    const { token } = await oauthApi.getToken(code);
+    localStorage.setItem('token', token);
+  };
```
질문) 저하고 수경님이 작업한 부분인데 여기서 로그인이 되었다는 상태로 변경해주는 작업이 필요할까요??

rolled-potatoes replied:
아뇨 저기서 '/'로 리다이렉트만 시키면 됩니다. 
아직 토큰으로 jwt 받아오는 부분이 없어서 비워둔 상태입니다.

changheedev commented on server/src/routes/oauth.js:
```diff
@@ -0,0 +1,41 @@
+import generateToken from '../libs/jwt';
+
+const router = express.Router();
+const clientURL = process.env.CLIENT_URL || 'http://localhost:4000';
+const codeMap = new Map();
+
```
앗.. 좋은 이름 생각해보기로 했다가 넘겨버렸네요 ㅋㅋㅋ

yejineee replied:
코드랑 유저의 맵이니깐 codeUserMap...?

changheedev commented on front/src/routes/user-store/user-context-actions.js:
```diff
@@ -0,0 +1,8 @@
+export default {
+  setUserInformation(information) {
+    return {
```
info는 자체로도 정보라는 의미를 가지니까 information으로 풀어쓰지 않으셔도 될 것 같습니다. 
너무 무분별하게 줄여쓰는걸 지양하기로 했었던거니까요 :)

changheedev commented on front/src/routes/user-store/user-context.js:
```diff
@@ -0,0 +1,28 @@
+const store = createContext(initialState);
+const { Provider } = store;
+
+const BaseProvider = ({ children }) => {
+  const [state, dispatch] = useReducer((userInformationState, action) => {
+    switch (action.type) {
+      case 'SET_USER_INFOMATION': {
+        const { info } = action.payload;
+        return { ...state, ...info, logged: true };
+      }
+      default: {
+        return userInformationState;
+      }
+    }
+  }, initialState);
```
이렇게 하면 좀 더 읽기 편한 코드가 될 것 같습니다.

```javascript
const userInfoReducer = (userInfoState, action) => {...};
const [state, dispatch] = useReducer(userInfoReducer, initialState);
```

rolled-potatoes replied:
콜백을 분리 할께요 ~

yejineee commented on front/src/routes/oauth/callback.js:
```diff
@@ -0,0 +1,20 @@
+      localStorage.setItem('token', token);
+    }
+    getToken().then(() => history.push('/'));
+  }, []);
```
비동기 처리는 es6의 async, await을 처리해주는게 더 깔끔하다고 생각합니다.
```js
await getToken();
history.push('/');
```

changheedev replied:
useEffect 콜백에서 async 를 사용할수가 없어서 then을 사용했어요

yejineee commented on front/src/App.js:
```diff
@@ -17,13 +22,32 @@ const GlobalStyled = createGlobalStyle`
-      <IssueListPage />
+      <Link to="login">로그인</Link>
+      <Switch>
+        <Route path="/mil" component={TempComponent} />
```
이 코드가 지금 필요한건가용...?

yejineee commented on front/src/routes/user-store/user-context-actions.js:
```diff
@@ -0,0 +1,8 @@
+      type: 'SET_USER_INFORMATION',
+      payload: information,
+    };
+  },
```
오 이 함수는 언제 필요한지 궁금하네용 🤩

rolled-potatoes replied:
아직은 구현안되어 있지만,
토큰으로 로그인 정보를 불러 왔을 때 동작할 리듀서 액션 입니다.

yejineee commented on front/src/routes/user-store/user-context.js:
```diff
@@ -0,0 +1,28 @@
+  const [state, dispatch] = useReducer((userInformationState, action) => {
+    switch (action.type) {
+      case 'SET_USER_INFOMATION': {
+        const { info } = action.payload;
```
action type은 다른 파일에서도 dispatch 부르면서 많이 사용하게 될 것 같은데, 객체로 생성해서 따로 관리하는 건 어떨까요?
오타를 내는 실수를 줄일 수 있을 것 같습니다...!
```js
//action-type.js
export default {
  SET_USER_INFORMATION: 'SET_USER_INFORMATION',
};

```

yejineee commented on front/src/routes/user-store/user-context.js:
```diff
@@ -0,0 +1,28 @@
+  return <Provider value={{ state, dispatch }}>{children}</Provider>;
+};
+
+export { store, BaseProvider };
```
import 하는 부분에서 as로 이름을 바꿔서 가져올 수는 있지만, export 할 때부터 명확한 이름을 정해서 주는건 어떨까용?
userState라던가.......허허

rolled-potatoes replied:
저기를 변경한다는게 잊어버렸네요.
수정하겠습니다.

rolled-potatoes committed:
메인 템플릿 분리 및 함수와 변수 네이밍 변경 (#80)

* Feat : 사용자 정보 관리 store 생성

사용자 정보를 전역으로 접근 할 수 있는 store를 생성 함.

* Feat : 프론트 라우팅 생성

* Feat : 프론트 oauth callback 함수 작성

* Feat : 프론트 라우팅 분리 및 user store 재정의

프론트에서 특정 URL 제외 하기위해 하드코딩 한 것을 필요 따라서 렌더링 될 수 있게
라우팅을 서브 라우팅으로 분리 하였음.

기존 BaseComponent가 적절한 네이밍 아니라고 판단하여 변경 함.

* Refactor : 메인 페이지 템플릿 분리 및 함수와 변수 네이밍 변경

App.js에서 모든 페이지 관리 할 때 파일이 비중해져서 따로 파일로 분리 하였음.

user store 네이밍이 base 되어 있는 것을 수정 함.

user store 리듀서에 타입을 실수를 방지하고자  별도 파일로 분리하여 관리 함.
